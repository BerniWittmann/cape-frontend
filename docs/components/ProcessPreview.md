# ProcessPreview

A Component to preview the svg of a process

## Props

<!-- @vuese:ProcessPreview:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|process|The Process object|`Object`|`true`|-|

<!-- @vuese:ProcessPreview:props:end -->


