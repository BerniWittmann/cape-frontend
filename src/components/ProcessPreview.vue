<template>
  <iframe v-bind:style="{ height: svgHeightAdjust }" :srcdoc="processSVG" class="svg-view margin-top"></iframe>
</template>

<script>

/*
 * @vuese
 * @group Components
 *
 * A Component to preview the svg of a process
 */
export default {
  name: 'ProcessPreview',
  props: {
    // The Process object
    process: {
      type: Object,
      required: true
    }
  },

  computed: {
    // @vuese
    // adjusts the height of the iframe to to either fit or be limited
    svgHeightAdjust() {
      if (!this.process.svg) return 'auto'
      const maxHeight = 500
      const heightPos = this.process.svg.indexOf('height="') + 8
      const height = parseInt(this.process.svg.substring(heightPos, this.process.svg.indexOf('"', heightPos)))
      return height <= maxHeight ? height + 25 + 'px' : maxHeight + 'px'
    },

    // @vuese
    // either returns the svg or a text if none available
    processSVG() {
      return !this.process.svg ? this.$t('process.error_svg_preview') : this.process.svg
    }
  }
}

</script>

<style lang="scss">
.svg-view {
  width: 100%;
  height: 100%;
  min-height: 100%;
  border: none;
}
</style>
