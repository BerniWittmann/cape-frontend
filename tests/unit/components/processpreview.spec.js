import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import ProcessPreview from '@/components/ProcessPreview.vue'

describe('Components', () => {
  describe('ProcessPreview', () => {
    let propsData
    let cmp

    beforeEach(() => {
      propsData = {
        process: {
          id: '142',
          name: 'My Process',
          svg: '<svg height="256px"><path d="1"></path></svg>'
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(ProcessPreview, {
        i18n,
        propsData
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
      expect(cmp.contains('iframe')).toBeTruthy()
    })

    it('renders the svg', () => {
      const iframe = cmp.find('iframe')
      expect(iframe.attributes('srcdoc')).toEqual(propsData.process.svg)
    })

    it('adjusts the height', () => {
      const iframe = cmp.find('iframe')
      expect(iframe.attributes('style')).toEqual('height: 281px;')
    })

    describe('it has no svg', () => {
      beforeEach(() => {
        propsData.process.svg = undefined
        render()
      })
      it('shows a text', () => {
        const iframe = cmp.find('iframe')
        expect(iframe.attributes('srcdoc')).toEqual('process.error_svg_preview')
      })

      it('adjusts the height', () => {
        const iframe = cmp.find('iframe')
        expect(iframe.attributes('style')).toEqual('height: auto;')
      })
    })

    describe('it has a very high svg', () => {
      beforeEach(() => {
        propsData.process.svg = '<svg height="512px"><path d="1"></path></svg>'
        render()
      })
      it('renders the svg', () => {
        const iframe = cmp.find('iframe')
        expect(iframe.attributes('srcdoc')).toEqual(propsData.process.svg)
      })

      it('adjusts the height', () => {
        const iframe = cmp.find('iframe')
        expect(iframe.attributes('style')).toEqual('height: 500px;')
      })
    })
  })
})
