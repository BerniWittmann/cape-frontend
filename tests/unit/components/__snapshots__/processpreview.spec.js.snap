// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ProcessPreview renders 1`] = `<iframe srcdoc="<svg height=&quot;256px&quot;><path d=&quot;1&quot;></path></svg>" class="svg-view margin-top" style="height: 281px;"></iframe>`;
